libraries/chain/controller.cpp

```diff
@@ -4394,145 +4293,126 @@ struct controller_impl {
       } FC_LOG_AND_RETHROW( )
    }
 
-   void maybe_switch_forks(const forked_callback_t& cb, const trx_meta_cache_lookup& trx_lookup) {
-      auto maybe_switch = [&](auto& forkdb) {
-         if (read_mode != db_read_mode::IRREVERSIBLE) {
-            auto pending_head = forkdb.head(include_root_t::yes);
-            if (chain_head.id() != pending_head->id()) {
-               dlog("switching forks on controller->maybe_switch_forks call");
-               controller::block_report br;
-               maybe_switch_forks(br, pending_head,
-                                  pending_head->is_valid() ? controller::block_status::validated
-                                                           : controller::block_status::complete,
-                                  cb, trx_lookup);
-            }
+   void apply_blocks(const forked_callback_t& cb, const trx_meta_cache_lookup& trx_lookup) {
+      try {
+         if( !irreversible_mode() ) {
+            maybe_apply_blocks( cb, trx_lookup );
+         } else {
+            log_irreversible();
+            transition_to_savanna_if_needed();
          }
-      };
-
-      fork_db.apply<void>(maybe_switch);
+      } FC_LOG_AND_RETHROW( )
    }
 
-   template<class BSP>
-   void maybe_switch_forks( controller::block_report& br, const BSP& new_head, controller::block_status s,
-                            const forked_callback_t& forked_cb, const trx_meta_cache_lookup& trx_lookup )
+   void maybe_apply_blocks( const forked_callback_t& forked_cb, const trx_meta_cache_lookup& trx_lookup )
    {
-      auto do_maybe_switch_forks = [&](auto& forkdb) {
-         if( new_head->header.previous == chain_head.id() ) {
-            try {
-               apply_block( br, new_head, s, trx_lookup );
-            } catch ( const std::exception& e ) {
-               dlog("Exception during apply block: ${e}", ("e", e.what()));
-               forkdb.remove( new_head->id() );
-               throw;
+      auto do_apply_blocks = [&](auto& forkdb) {
+         auto new_head = forkdb.head(); // use best head
+         if (!new_head)
+            return; // nothing to do, forkdb at root
+         auto branches = forkdb.fetch_branch_from( new_head->id(), chain_head.id() );
```

Comment:
I think this might be clearer:

```suggestion
         auto [new_head_branch, old_head_branch] = forkdb.fetch_branch_from( new_head->id(), chain_head.id() );
```

Reply:
Much cleaner.